### ISSUE TITLE
Exceptions populate as generic "C++ exception"; exception handlers and xpcall callbacks not called.

### ISSUE THREAD

**Issue by NullCascade** [ThePhD/sol2#1038]
**Compiler**: Visual Studio 2019 v16.7.5
**Lua sku**: LuaJIT 2.1.0-beta3 (x86)

I'm having an issue where C++ exceptions are not being handled correctly in a couple of places. I have not changed LuaJIT lately, though did make a big update to change from sol2 to sol3 recently. As far as I can tell, running:
```lua
local status, result = xpcall(callback, onEventError, payload)
```
`onEventError` will never actually be called.

Similarly, [exceptions example](https://sol2.readthedocs.io/en/latest/exceptions.html) provides some oddities, in that the only console output is:
```
=== exception_handler ===
C++ exception
```
rather than having the actual `.what()` of the exception.

This is with both `SOL_ALL_SAFETIES_ON` and `SOL_EXCEPTIONS_SAFE_PROPAGATION` defined.

I'm not yet entirely convinced that this is a sol-specific issue, as I don't think sol does anything related to `xpcall`. However, I would appreciate any wisdom that can be found here. Again, this same environment used to work just fine for me.

**Comment by jdbool** [comment]
Updated from 3.2.1 to 3.2.2, and the `.what()` is now the generic "C++ exception" instead of the useful info it had before.

**Comment by ThePhD** [comment]
Yeah this is an ongoing issue and is related to the other (#965) problem...

I need a better way of handling exceptions throughout, and then ways to configure how that is handled better for the end-user to get the behavior they want in all cases. There is some overlap from nested functions and the like, though, so I need to be careful.....
